File: Contests/CodeforcesPython/Educational Codeforces Round 159 Rated for Div. 2/E_Collapsing_Strings/E_Collapsing_Strings.py
import sys

# region fastio
input = lambda: sys.stdin.readline().rstrip()
sint = lambda: int(input())
mint = lambda: map(int, input().split())
ints = lambda: list(map(int, input().split()))
# endregion fastio

# MOD = 998_244_353
# MOD = 10 ** 9 + 7
# DIR4 = ((-1, 0), (0, 1), (1, 0), (0, -1)) #URDL
# DIR8 = ((-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1))

class Trie:
    def __init__(self):
        self.to = [[-1] for _ in range(26)]
        self.cnt = [0]

    def insert(self, word: str) -> None:
        cur = 0
        for c in word:
            c = ord(c) - 97
            if self.to[c][cur] == -1:
                self.to[c][cur] = len(self.cnt)
                for i in range(26):
                    self.to[i].append(-1)
                self.cnt.append(0)
            cur = self.to[c][cur]
            self.cnt[cur] += 1

    def search(self, word: str) -> int:
        res = 0
        cur = 0
        for c in word:
            c = ord(c) - 97
            if self.to[c][cur] == -1:
                return res
            cur = self.to[c][cur]
            res += self.cnt[cur]
        return res

def solve() -> None:
    n = sint()
    ans = tot = 0
    s = []
    trie = Trie()
    for _ in range(n):
        s.append(input())
        tot += len(s[-1])
        trie.insert(s[-1][::-1])
    for word in s:
        cnt = trie.search(word)
        ans += tot + len(word) * n - cnt * 2
    print(ans)

solve()

File: Contests/CodeforcesPython/Educational Codeforces Round 159 Rated for Div. 2/E_Collapsing_Strings/E_Collapsing_Strings.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
)

func solve(_r io.Reader, _w io.Writer) {
	in := bufio.NewReader(_r)
	out := bufio.NewWriter(_w)
	defer out.Flush()

	var n int
	Fscan(in, &n)
	var ans, tot int
	s := make([]string, n)
	trie := newTrie()
	for i := range s {
		Fscan(in, &s[i])
		tot += len(s[i])
		trie.insert((s[i]))
	}
	for i := range s {
		cnt := trie.search(s[i])
		ans += tot + len(s[i])*n - cnt*2
	}
	Fprintln(out, ans)
}

func main() { solve(os.Stdin, os.Stdout) }

type Trie struct {
	children map[byte]*Trie
	cnt      int
}

func newTrie() *Trie {
	m := map[byte]*Trie{}
	return &Trie{children: m}
}
func (this *Trie) insert(w string) {
	node := this
	for i := len(w) - 1; i >= 0; i-- {
		if _, ok := node.children[w[i]]; !ok {
			node.children[w[i]] = newTrie()
		}
		node, _ = node.children[w[i]]
		node.cnt++
	}
}
func (this *Trie) search(w string) int {
	node := this
	res := 0
	for i := range w {
		if _, ok := node.children[w[i]]; !ok {
			return res
		}
		node, _ = node.children[w[i]]
		res += node.cnt
	}
	return res
}
